rfcs/hse/0004-kvdb-kvs-introspection-apis.md

```diff
@@ -0,0 +1,154 @@
+
+#### C APIs
+
+The following C APIs will be added behind the experimental wall:
+
+```c
+/* Should this API return an error in the case of a NULL KVDB?
+ *
+ * Returns a pointer to the kvdb_home member in ikvdb struct.
+ */
+const char *
+hse_kvdb_home_get(const struct hse_kvdb *kvdb);
+
+/* Should this API return an error in the case of a NULL KVDB?
+ *
+ * Returns a pointer to the name member in ikvs struct.
+ */
+const char *
+hse_kvs_name_get(const struct hse_kvs *kvs);
+
+/* Returns a JSON representation of the parameter value, ie the value in
+ * key=value.
+ */
+hse_err_t
+hse_kvdb_param_get(const struct hse_kvdb *kvdb, const char *param, size_t buf_len, char *buf);
+
+/* Returns a JSON representation of the parameter value, ie the value in
+ * key=value.
+ */
+hse_err_t
+hse_kvs_param_get(const struct hse_kvs *kvs, const char *param, size_t buf_len, char *buf);
+
+enum hse_mclass {
+  HSE_MCLASS_CAPACITY,
+  HSE_MCLASS_STAGING,
+};
+
+struct hse_kvdb_storage_info {
+  uint64_t total_bytes;     /**< total space in the file-system containing this kvdb */
+  uint64_t available_bytes; /**< available space in the file-system containing this kvdb */
+  uint64_t allocated_bytes; /**< allocated storage space for a kvdb */
+  uint64_t used_bytes;      /**< used storage space for a kvdb */
+  char     path[PATH_MAX];  /**< path to media class */
+};
+
+/* Populates the above storage info struct with given information.
+ *
+ * Con: Since we only track data owned by each media class, we drop the storage
+ * statistics about the kvdb.meta file, but I don't see that as being super
+ * important.
+ */
+hse_err_t
+hse_kvdb_storage_info_get(const struct hse_kvdb *kvdb, struct hse_kvdb_storage_info *info);
+```
+
+#### REST APIs
+
+One issue we have currently that is worth solving in this RFC is that our REST
+API assumes there is only one KVDB. We seem to have settled on the idea we want
+to support more than one KVDB in the future. Now is the perfect time to retrofit
+the REST API.
+
+One issue with name-less KVDBs is that file paths look like parts of a REST
+path, which is not good for easy usage. Typically in a well-planned REST route
+tree, you would have something like `/kvdb/<specific kvdb>`, where `/kvdb` is
+the parent route of all KVDBs. In our current, name-less KVDB situation this
+would look like `/kvdb/path/to/kvdb` where `path/to/kvdb` is the KVDB home. This
+makes it impossible to parse REST paths correctly. What we can do instead is the
+following:
+
+- Add a new KVDB rparam: `alias`
+  - The value of `alias` would default to `0` for the first KVDB opened, `1` for
+    the next KVDB opened and so on. In the event, KVDB 0 is closed, and KVDB 2
+    is opened, KVDB 2 would take over the alias 0. Alex has suggested using an
+    always incrementing alias.
+  - The rparam could also be called `socket.alias`, but I have a feeling the
+    alias could come in handy in other sections of our code, and it doesn't
+    hamstring us.
+
```

Comment:
Alternative: Rather than a KVDB rparam, just use the default approach.  And I agree with Alex about just always incrementing the identifier, seems safer that /kvdb/0 never changes (though it can be closed) during the process lifetime.

Overall I'm not a fan of alias here or in a future use.  We looked at several options before converging on the KVDB home directory approach, and alias feels counter to that decision.

Reply:
I feel like it is entirely complimentary. In a multi-kvdb application, someone would name their kvdb variable `<alias>_db` or something of that sort because paths are worthless for a variable name.

Reply:
I am not entirely attached to the alias idea, but I definitely think it solves the problem of the REST routes better than an autoincrementing numeric identifier.

Reply:
From the HSE API this is a non-issue because I can use any variable name to hold the result of hse_kvdb_open().  This is primarily a REST issue, and the REST API will only have limited uses.  So if we feel strongly about the alias option for REST, then let's put this under the socket key.  That feels like a reasonable middle ground to land on.

Reply:
I guess the beauty is that we can always change this in a backwards compatible way. I say let's forget about the notion of a formalized alias for now and just use the auto-incrementing numeric identifier for now without the rparam.

Reply:
Not to keep beating this poor horse, but something that might be more REST friendly w/o specifying an rparam alias is to use /kvdb/last-component-of-home-directory-path, and in the rare case where a process opens 2+ KVDB with same final home directory path component disambiguate them in some well-defined way.  E.g., /kvdb/mykvdb, /kvdb/mykvdb-2.
In this rare case the app would have to get some additional information to fully distinguish them, like the full home directory paths.

Reply:
I'll let Alex provide his input, but I think there is value in not needing to know the what the last part of the KVDB path is and whether or not it clashes with another open KVDB. 0 is always easy to remember as the default for when there exists only one KVDB.

Reply:
I'm good with this as is.